File: app/src/main/java/com/mamba/picme/domain/person/RelationPredicate.kt
package com.mamba.picme.domain.person

/**
 * 人物关系谓词（封闭枚举）—— 人物关系图谱边的类型
 *
 * 每个取值带中/英/日三语标签，供命名对话框下拉与聊天工具展示使用；
 * 数据库存储枚举名（[name]），通过 [fromStored] 还原。
 */
enum class RelationPredicate(
    val labelZh: String,
    val labelEn: String,
    val labelJa: String
) {
    SPOUSE(labelZh = "配偶", labelEn = "Spouse", labelJa = "配偶者"),
    PARTNER(labelZh = "恋人", labelEn = "Partner", labelJa = "恋人"),
    CHILD(labelZh = "孩子", labelEn = "Child", labelJa = "子供"),
    SON(labelZh = "儿子", labelEn = "Son", labelJa = "息子"),
    DAUGHTER(labelZh = "女儿", labelEn = "Daughter", labelJa = "娘"),
    PARENT(labelZh = "父母", labelEn = "Parent", labelJa = "親"),
    FATHER(labelZh = "爸爸", labelEn = "Father", labelJa = "父"),
    MOTHER(labelZh = "妈妈", labelEn = "Mother", labelJa = "母"),
    SIBLING(labelZh = "兄弟姐妹", labelEn = "Sibling", labelJa = "兄弟姉妹"),
    ELDER_BROTHER(labelZh = "哥哥", labelEn = "Elder brother", labelJa = "兄"),
    ELDER_SISTER(labelZh = "姐姐", labelEn = "Elder sister", labelJa = "姉"),
    YOUNGER_BROTHER(labelZh = "弟弟", labelEn = "Younger brother", labelJa = "弟"),
    YOUNGER_SISTER(labelZh = "妹妹", labelEn = "Younger sister", labelJa = "妹"),
    GRANDPARENT(labelZh = "祖辈", labelEn = "Grandparent", labelJa = "祖父母"),
    GRANDFATHER(labelZh = "爷爷", labelEn = "Grandfather", labelJa = "祖父"),
    GRANDMOTHER(labelZh = "奶奶", labelEn = "Grandmother", labelJa = "祖母"),
    GRANDCHILD(labelZh = "孙辈", labelEn = "Grandchild", labelJa = "孫"),
    OTHER_FAMILY(labelZh = "其他亲属", labelEn = "Other family", labelJa = "その他の親族"),
    FRIEND(labelZh = "朋友", labelEn = "Friend", labelJa = "友人"),
    CLASSMATE(labelZh = "同学", labelEn = "Classmate", labelJa = "クラスメート"),
    COLLEAGUE(labelZh = "同事", labelEn = "Colleague", labelJa = "同僚"),
    OTHER(labelZh = "其他", labelEn = "Other", labelJa = "その他");

    companion object {
        /** 数据库字符串 → 枚举；无法识别返回 null（调用方决定降级策略） */
        fun fromStored(stored: String): RelationPredicate? =
            values().firstOrNull { predicate -> predicate.name == stored }
    }
}

/**
 * 关系声明来源（封闭枚举）—— 数据库存储枚举名
 */
enum class RelationSource {
    /** 相册人物重命名对话框 */
    RENAME_DIALOG,

    /** 聊天声明工具（"记住 X 是我 Y"） */
    CHAT_DECLARATION;

    companion object {
        fun fromStored(stored: String): RelationSource? =
            values().firstOrNull { source -> source.name == stored }
    }
}

File: iosApp/PoLang/Platform/Search/RelationPredicate.swift
import Foundation

// MARK: - 人物关系谓词（双端契约 SSOT: contracts.md §7.4）
//
// 逐字对齐 shared `RelationPredicate.kt:13-47`：封闭枚举 + 中/英/日标签。
// 数据库存枚举名字符串（rawValue），`fromStored` 按 name 精确还原（无法识别 → nil，
// 调用方决定降级策略）。标签仅供命名 UI 展示，搜索不消费。

/// 人物关系谓词（封闭枚举）——人物关系图谱边的类型。
enum RelationPredicate: String, CaseIterable, Sendable {
    case spouse = "SPOUSE"
    case partner = "PARTNER"
    case child = "CHILD"
    case son = "SON"
    case daughter = "DAUGHTER"
    case parent = "PARENT"
    case father = "FATHER"
    case mother = "MOTHER"
    case sibling = "SIBLING"
    case elderBrother = "ELDER_BROTHER"
    case elderSister = "ELDER_SISTER"
    case youngerBrother = "YOUNGER_BROTHER"
    case youngerSister = "YOUNGER_SISTER"
    case grandparent = "GRANDPARENT"
    case grandfather = "GRANDFATHER"
    case grandmother = "GRANDMOTHER"
    case grandchild = "GRANDCHILD"
    case otherFamily = "OTHER_FAMILY"
    case friend = "FRIEND"
    case classmate = "CLASSMATE"
    case colleague = "COLLEAGUE"
    case idol = "IDOL"
    case other = "OTHER"

    /// 中文标签（RelationPredicate.kt labelZh）
    var labelZh: String {
        switch self {
        case .spouse: return "配偶"
        case .partner: return "恋人"
        case .child: return "孩子"
        case .son: return "儿子"
        case .daughter: return "女儿"
        case .parent: return "父母"
        case .father: return "爸爸"
        case .mother: return "妈妈"
        case .sibling: return "兄弟姐妹"
        case .elderBrother: return "哥哥"
        case .elderSister: return "姐姐"
        case .youngerBrother: return "弟弟"
        case .youngerSister: return "妹妹"
        case .grandparent: return "祖辈"
        case .grandfather: return "爷爷"
        case .grandmother: return "奶奶"
        case .grandchild: return "孙辈"
        case .otherFamily: return "其他亲属"
        case .friend: return "朋友"
        case .classmate: return "同学"
        case .colleague: return "同事"
        case .idol: return "偶像"
        case .other: return "其他"
        }
    }

    /// 英文标签（RelationPredicate.kt labelEn）
    var labelEn: String {
        switch self {
        case .spouse: return "Spouse"
        case .partner: return "Partner"
        case .child: return "Child"
        case .son: return "Son"
        case .daughter: return "Daughter"
        case .parent: return "Parent"
        case .father: return "Father"
        case .mother: return "Mother"
        case .sibling: return "Sibling"
        case .elderBrother: return "Elder brother"
        case .elderSister: return "Elder sister"
        case .youngerBrother: return "Younger brother"
        case .youngerSister: return "Younger sister"
        case .grandparent: return "Grandparent"
        case .grandfather: return "Grandfather"
        case .grandmother: return "Grandmother"
        case .grandchild: return "Grandchild"
        case .otherFamily: return "Other family"
        case .friend: return "Friend"
        case .classmate: return "Classmate"
        case .colleague: return "Colleague"
        case .idol: return "Idol"
        case .other: return "Other"
        }
    }

    /// 日文标签（RelationPredicate.kt labelJa）
    var labelJa: String {
        switch self {
        case .spouse: return "配偶者"
        case .partner: return "恋人"
        case .child: return "子供"
        case .son: return "息子"
        case .daughter: return "娘"
        case .parent: return "親"
        case .father: return "父"
        case .mother: return "母"
        case .sibling: return "兄弟姉妹"
        case .elderBrother: return "兄"
        case .elderSister: return "姉"
        case .youngerBrother: return "弟"
        case .youngerSister: return "妹"
        case .grandparent: return "祖父母"
        case .grandfather: return "祖父"
        case .grandmother: return "祖母"
        case .grandchild: return "孫"
        case .otherFamily: return "その他の親族"
        case .friend: return "友人"
        case .classmate: return "クラスメート"
        case .colleague: return "同僚"
        case .idol: return "アイドル"
        case .other: return "その他"
        }
    }

    /// 数据库字符串 → 枚举；无法识别返回 nil（对齐 `RelationPredicate.fromStored`）。
    static func fromStored(_ stored: String) -> RelationPredicate? {
        RelationPredicate(rawValue: stored)
    }
}
